[pytorch] add fused compute_grad_weight kernel when segments are few but large

yaoyj11 opened the pull request:
Current 2-step sum and scatter approach is not efficient when the number of segments is small but each segment is large. For example, we observed in our prod job that there is a sum_and_scatter kernel taking 40ms, but the launch grid size is 100 on Blackwell GPU (148 SMs). 1/3 of the SMs are idle why the other SMs are overloaded. And the kernel is bottlenecked by the largest segment in our case.
To improve the efficiency in this case, we add a fused atomic add kernel with grid size proportional to max_partial_segment instead of max_segments, giving much better parallelism when segments are few but large

Test
1. Added unit test to cover the atomic accumulate kernel path. The large atol and rtol thresholds in the tests are needed for both original path and new atomic accmulate path.

2. Verified numerical parity and perf improvement with e2e workflow.
Baseline: fire-renqincai-20251210-2133-a14d8be3, sum_and_scatter kernel takes about 40ms per iteration.
After optimization: fire-yjyao-20251214-1731-faeab2e0, reduced to about 6ms.

4% qps gain, NE on par

Differential Revision: D89141291

jcaip commented:
cc @syed-ahmed @eqy @Aidyn-A this looks okay to me, any objections to merging?

eqy commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -153,6 +153,79 @@ __global__ void compute_grad_weight(
+// This eliminates the serialization bottleneck when num_of_segments is small.
+// Each partial segment atomically adds its contribution directly to grad_weight.
+template <typename scalar_t, typename index_t>
+__global__ void compute_grad_weight_fused(
```
nit: maybe we could go with a more descriptive kernel name here (imagine seeing this kernel name e.g., in nsys nvprof)

eqy commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -153,6 +153,79 @@ __global__ void compute_grad_weight(
+  const int64_t target_row = sorted_indices[segment_offsets[segment_id]];
+
+  if (target_row != padding_idx) {
+    gpuAtomicAddNoReturn(
```
Are we now using atomics in a path that didn't use them, do we need to check if we are doing things in deterministic mode?

eqy reviewed (commented):
Have you confirmed that existing tests are exercising the fused path?

yaoyj11 commented:
The two failing checks are due to timeouts and not related with this change. @eqy I added new unit tests for the new path and verified that they are used during test locally. Could you help take another look? Thanks.

yaoyj11 commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -153,6 +153,79 @@ __global__ void compute_grad_weight(
+  if (id < num_of_segments) {
+    index_t idx = partials_per_segment_offset[id];
+    const index_t num_partials = partials_per_segment[id];
+    for (int64_t i = 0; i < num_partials; ++i) {
```
@eqy @jcaip Could you help taking another review? I went through the code again and hopefully there is no more misuse of data types. I appreciate your help.

jcaip commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -285,59 +358,107 @@ Tensor embedding_backward_cuda_kernel(
+    // Note: Fused kernel uses atomics which are non-deterministic, so we skip it
+    // when deterministic algorithms are requested
+    const int64_t sum_scatter_grid = ceil_div(max_segments * stride_warped, block);
+    const int64_t min_blocks_for_good_occupancy = 1024;  // Tunable threshold
```
Are we okay with this threshold being hard coded?
Does this threshold depend on what GPU hardware you have?

jcaip commented on test/nn/test_embedding.py:
```diff
@@ -653,6 +653,129 @@ def gen_2D_indices_from_1D(
+        2. Few unique indices (small num_of_segments) resulting in poor parallelism
+           for the original sum_and_scatter kernel
+        3. Many repeated indices creating many partial segments
+        4. In non-deterministic mode
```
nit: can we add `torch.use_deterministic_algorithms(False)` to make it a bit clearer that we're running in non-determinstic mode.

jcaip reviewed (commented):
cc @yaoyj11 thanks, had a couple more comments. 

can you also update the PR body a bit more with summary, test plan, etc, just so it's a bit more understandable. It would be great if you can link benchmarks as well if they are not internal.

jcaip reviewed (approved):
approving to unblock, but please do update the PR body before you land.

yaoyj11 commented:
> approving to unblock, but please do update the PR body before you land.

Thanks for reviewing! @eqy Could you help review this PR again as well? Approval from maintainers are required to merge the PR.

yaoyj11 committed:
[pytorch] add fused compute_grad_weight kernel when segments are few but large (#170589)

Summary:
Current 2-step sum and scatter approach is not efficient when the number of segments is small but each segment is large. For example, we observed in our prod job that there is a sum_and_scatter kernel taking 40ms, but the launch grid size is 100 on Blackwell GPU (148 SMs). 1/3 of the SMs are idle why the other SMs are overloaded. And the kernel is bottlenecked by the largest segment in our case.
To improve the efficiency in this case, we add a fused atomic add kernel with grid size proportional to max_partial_segment instead of max_segments, giving much better parallelism when segments are few but large

Test
1. Added unit test to cover the atomic accumulate kernel path. The large atol and rtol thresholds in the tests are needed for both original path and new atomic accmulate path.

2. Verified numerical parity and perf improvement with e2e workflow.
Baseline: fire-renqincai-20251210-2133-a14d8be3, sum_and_scatter kernel takes about 40ms per iteration.
After optimization: fire-yjyao-20251214-1731-faeab2e0, reduced to about 6ms.

4% qps gain, NE on par


Reviewed By: jcaip

Differential Revision: D89141291

Pulled By: yaoyj11
